packages/queued-request-controller/src/QueuedRequestController.ts

```diff
@@ -263,30 +282,61 @@ export class QueuedRequestController extends BaseController<
    */
   async #processNextBatch() {
     const firstRequest = this.#requestQueue.shift() as QueuedRequest;
-    this.#originOfCurrentBatch = firstRequest.origin;
-    this.#networkClientIdOfCurrentBatch = firstRequest.networkClientId;
-    const batch = [firstRequest.processRequest];
+    this.#originOfCurrentBatch = firstRequest.request.origin;
+    this.#networkClientIdOfCurrentBatch = firstRequest.request.networkClientId;
+    const batch = [firstRequest];
+
+    let networkSwitchError: unknown;
+    try {
+      // If globally selected network is different from origin selected network,
+      // switch network before processing batch
+      await this.#switchNetworkIfNecessary(
+        firstRequest.request.networkClientId,
+      );
+    } catch (error: unknown) {
+      networkSwitchError = error;
+    }
+
+    // If the first request might switch the network, process the request by
+    // itself. If the request does change the network, clear the queue for the
+    // origin since it any remaining requests are now invalidated
+    if (this.#canRequestSwitchNetworkWithoutApproval(firstRequest.request)) {
+      // This hack prevents the next batch from being processed
+      // after this request returns. This is necessary because
+      // we may need to flush the queue before the next set of requests
+      // are batched and processed, which we cannot do without blocking
+      // the queue from continuing by artificially increasing the processing
+      // request count
+      this.#processingRequestCount += 1;
+      firstRequest.processRequest(networkSwitchError);
+      this.#updateQueuedRequestCount();
+      await firstRequest.processedPromise;
+      this.#processingRequestCount -= 1;
+      const { selectedNetworkClientId } = this.messagingSystem.call(
+        'NetworkController:getState',
+      );
+      if (this.#networkClientIdOfCurrentBatch !== selectedNetworkClientId) {
+        this.#flushQueueForOrigin(this.#originOfCurrentBatch);
+      }
+      this.#processNextBatchIfReady();
```

Comment:
Why is this call needed? The next batch should still get triggered the same way as before, when the end of `enqueueRequest` is hit for the switch/add Ethereum chain method call. It seems like it's getting triggered twice here now.

Reply:
needed because we prevent the add/switch request from triggering this at the end of enqueueRequest now via the ` this.#processingRequestCount += 1;` and ` this.#processingRequestCount -= 1;` hack above. 

Hacky guard is needed because without it `enqueueRequest` for the wallet_add/switch request will trigger `processNextBatch` before we are able to check if the network changed and call ` this.#flushQueueForOrigin(this.#originOfCurrentBatch);` above this line

Reply:
Right... got my order of operations wrong. I thought that this would run before the dequeued promise had resolved. But no, the dequeued promise would resolve first, and never trigger the next batch. Understood, makes perfect sense.

Maybe worth adding a comment here to explain why this is needed?